File: frontend/src/lib/api/globalSearch.ts
/**
 * Global Search API client.
 *
 * Provides API methods for searching across all matters the user has access to.
 * Story 14.11: Global Search RAG Wiring
 */

import { api } from './client';

/**
 * Search result item from the global search API.
 * Matches the frontend SearchResult interface in GlobalSearch.tsx.
 */
export interface SearchResult {
  id: string;
  type: 'matter' | 'document';
  title: string;
  matterId: string;
  matterTitle: string;
  matchedContent: string;
}

/**
 * Global search response metadata.
 */
interface GlobalSearchMeta {
  query: string;
  total: number;
}

/**
 * Global search API response.
 */
interface GlobalSearchResponse {
  data: SearchResult[];
  meta: GlobalSearchMeta;
}

/**
 * Search across all matters the user has access to.
 *
 * Uses hybrid search (BM25 + semantic) with cross-matter RRF fusion.
 * Results include both matter title matches and document content matches.
 *
 * @param query - Search query (minimum 2 characters)
 * @param limit - Maximum results to return (default 20, max 50)
 * @returns Array of search results
 *
 * @example
 * const results = await globalSearch('contract breach');
 * // Returns matters and documents matching "contract breach"
 */
export async function globalSearch(
  query: string,
  limit = 20
): Promise<SearchResult[]> {
  const params = new URLSearchParams({
    q: query,
    limit: limit.toString(),
  });

  const response = await api.get<GlobalSearchResponse>(
    `/api/search?${params.toString()}`
  );

  return response.data;
}

File: backend/app/models/global_search.py
"""Global Search models for cross-matter search API.

Defines request/response models for the global search endpoint.
This endpoint searches across ALL matters the user has access to.
"""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class GlobalSearchResultItem(BaseModel):
    """Single global search result item.

    Uses camelCase aliases to match frontend SearchResult interface
    from GlobalSearch.tsx.
    """

    model_config = ConfigDict(populate_by_name=True)

    id: str = Field(..., description="Result UUID (chunk ID or matter ID)")
    type: Literal["matter", "document"] = Field(
        ..., description="Result type: 'matter' or 'document'"
    )
    title: str = Field(..., description="Result title (document name or matter title)")
    matter_id: str = Field(
        ..., alias="matterId", description="Matter UUID this result belongs to"
    )
    matter_title: str = Field(
        ..., alias="matterTitle", description="Title of the matter"
    )
    matched_content: str = Field(
        ...,
        alias="matchedContent",
        description="Snippet of matched content (50-100 chars)",
    )


class GlobalSearchMeta(BaseModel):
    """Metadata about the global search results."""

    query: str = Field(..., description="Original search query")
    total: int = Field(..., description="Total number of results returned")


class GlobalSearchResponse(BaseModel):
    """Response model for global search endpoint.

    Follows project API response format with data and meta.
    """

    data: list[GlobalSearchResultItem] = Field(..., description="Search results")
    meta: GlobalSearchMeta = Field(..., description="Search metadata")


class GlobalSearchErrorDetail(BaseModel):
    """Error detail model for global search errors."""

    code: str = Field(..., description="Error code")
    message: str = Field(..., description="Human-readable error message")
    details: dict = Field(default_factory=dict, description="Additional error details")


class GlobalSearchErrorResponse(BaseModel):
    """Error response model for global search endpoint."""

    error: GlobalSearchErrorDetail = Field(..., description="Error information")
